Implement loop to wrap array functions

efeichen opened the pull request:
This PR complements #374. The loop here checks if the `arrays` metadata is present and wraps the Mavo function if so.

The block inside `else` allows certain arguments of a function to accept arrays, as specified by the array of argument indices in an `arrays` property. The block also lets multi-argument functions like `between`, `random`, and `replace` to accept array arguments.

When we want one of the arguments of a **binary function** to accept arrays, however, fallback metadata like `default` and `unary` won't be utilized since `binaryOperation()` is not called. There won't be any uneven arrays in this case but I'm still wondering if this is acceptable?

Also, a caveat I see is that default parameters will cause `function.length` to return incorrectly, which could make wrapping inaccurate if `arrays: true`. For example, `replace()` has 2 default parameters so if one wants all args to accept arrays, he/she has to specify `arrays: [0,1,2,3]` or else binary operations will be performed on `replace`.

efeichen commented on src/functions.js:
```diff
@@ -513,6 +513,34 @@ var _ = Mavo.Functions = {
+		}
+	}
+});
+
```
Yes, this is to support the use of multi-value arguments for functions that ask for multiple arguments. I'll look into reimplementing with `binaryOperation()` instead (and more comments). 😃

efeichen replied:
For functions that are not associative, I'm wondering if using `binaryOperation()` multiple times would still work (e.g. `digits(1, 2, 3.141)`)? Multi-arg functions like `max` and `min` are out of the question after our previous discussions. But I'm concerned about these non-associative ones that could benefit from having muti-valued arguments.

LeaVerou replied:
It should, assuming it's used in a way that ends up calling the function with single-valued parameters at all times.

efeichen commented on src/functions.js:
```diff
@@ -513,6 +513,34 @@ var _ = Mavo.Functions = {
+		else if (FN.length === 2 && ARRAYS === true || ARRAYS.length === 2) {
+			Mavo.Functions[property] = (a, b) => Mavo.Script.binaryOperation(a, b, FN);
+		}
+		else {
```
I think `if (ARRAYS)` on line 520 is checking whether `ARRAYS` is undefined or not, and skips the wrapping if so. Am *I* missing something?

LeaVerou replied:
Nope, I was! You're right :)

efeichen commented on src/functions.js:
```diff
@@ -513,6 +513,34 @@ var _ = Mavo.Functions = {
+					result[i] = FN(...args.map(n => acceptedArrays.includes(n) ? n[i] : n));
+				}
+
+				return acceptedArrays.length !== 0 ? result : result[0];
```
`acceptedArrays.length === 0` means that all the arguments that accept arrays are not arrays. If this condition is not checked the result is always wrapped by an array, which is not ideal when no multi-valued arguments are present.

efeichen commented on src/functions.js:
```diff
@@ -513,6 +513,34 @@ var _ = Mavo.Functions = {
 
+Object.getOwnPropertyNames(Mavo.Functions).forEach(property => {
+	const FN = Mavo.Functions[property];
+	const ARRAYS = FN.arrays;
```
After some thought, I do like `multiValued` and I can't think of any other 2-word names that sound better than this.

Nevertheless, do you think it's necessary for us to document or comment on the purpose of these meta properties like `multiValued`, `unary`, and `rightScalar`, after we finalize these PRs?

LeaVerou replied:
Absolutely!

efeichen commented:
Addressed your comment about asynchronous wrapping and I hope I did it right with the `dependencies[0].then()`. Besides adding some comments, I also added another loop that wraps all unary Math functions (which I think is somewhat relevant to this PR?)

That code block in `else{}` is unmodified yet because I still have a question which I left you on Gitter. :)

LeaVerou commented:
Hey, sorry, I had missed your Gitter message. I'll think about it, however I think in the interest of getting these changes merged as soon as possible, we can limit the number of array arguments to 2 (either the first two, or the first two specified) and use `binaryOperation()` as is. This by itself will cover *most* use cases, and it will be easy to merge quickly. Then we can work on extending it as separate functionality, but (IMHO) it's a bad idea to let PRs stay open for so long and stagnate, it's better to cut down the functionality to what can be merged asap, and leave the rest as a future enhancement. The longer a PR is open, the more everyone forgets about its details and it becomes progressively harder to merge it.

I'm confident we can adapt `binaryOperation()` for more array arguments too, but need to think what's the best way to proceed with this.

efeichen commented:
Ya, I agree! Definitely will take note of the size in the future after having pushed a PR that stagnated for so long!

LeaVerou commented:
Right now, this requires specifying the arguments in `multiValued` explicitly for any function with more than 2 arguments. What if we just default to the first 2 arguments in those cases, if `multiValued` is true?

efeichen commented:
Not sure if the new commit is the change you're looking for? Did you mean that I should revert the change that removed the `else{}` block and revise it instead so that it defaults to the first 2 args?

LeaVerou commented:
Looks good! Have you tested it? Does it still work?

LeaVerou commented on src/functions.js:
```diff
@@ -513,6 +513,28 @@ var _ = Mavo.Functions = {
+ * After plugins are loaded, enable
+ * multi-valued arguments of Mavo and Math functions
+ */
+requestAnimationFrame(() => {
```
If you're gonna wrap the code in a Promise#then(), you don't need this, it's executed after everything else anyway (even if the promise is resolved at the time you called the `then()`!).

LeaVerou commented on src/functions.js:
```diff
@@ -513,6 +513,28 @@ var _ = Mavo.Functions = {
+	Mavo.dependencies[0].then(() => {
+		Object.getOwnPropertyNames(Mavo.Functions).forEach(property => {
+			const FN = Mavo.Functions[property];
+			const ARRAYS = FN.multiValued;
```
It would be useful to enable this on some functions, in the same PR. Otherwise once this is merged, it just adds extra code with no immediate benefit.

- Are there any functions that can be simplified with this? I.e. functions that currently take care of arrays manually and we can shorten them by just setting this flag on them?
- There are two ways to enable this flag on the functions that need it:
	- In the function definition, changing `functionName: function(){...}` to `functionName: $.extend(function(){...}, {multiValued: true})`. This may get a little repetitive for many functions, but is very explicit (an the only solution for values other than `true`). I also think it might be *more* DRY than other solutions, despite the repetitiveness, as we don't need to maintain info about functions in two places.
	- After all functions are defined, in a loop that goes over a list of names and setting `multiValued` to `true`. This may be good for handling them in bulk, but we do need to maintain names in two places (and change them in two places) and we still need to use the first way for values other than `true`.

efeichen replied:
> Are there any functions that can be simplified with this? I.e. functions that currently take care of arrays manually and we can shorten them by just setting this flag on them?

There are quite a lot of functions that call `binaryOperation()` so we could definitely simply them.

> There are two ways to enable this flag on the functions that need it:

I do like the first way more than the latter, and I believe it's what we originally came up with.

LeaVerou commented:
You can pass in a function to binaryOperation which wraps the actual function, where all arguments are fixed and the only two arguments are the ones that are arrays. That's what I was suggesting earlier, not to have a PR that doesn't work!

E.g. something like `(a, b) => actualFunction(1, someVariable, a, "foo", b)` where the multivalued arguments are the 3rd and the 5th, and everything else is fixed to their single-valued values. Does ths make sense?

efeichen commented:
I may have misinterpreted your previous comment. But thank you so much for clarifying. It makes a lot more sense now.

efeichen commented:
Everything works locally now. I removed the `argument.length >= 2` constraint to support functions with all default parameters (e.g. `random` which is multi-arg but has `arguments.length` of 0).

Added `...FN` to import `default` metadata, but it looks like support for object spread operator is not supported here? Should I explicitly specify `FN.default, FN.rightDefault, FN.unary ... etc.` instead?

LeaVerou commented:
That's a nice usage of the object spread operator. It is supported on recent Chrome and Firefox, but not Edge/IE. That's not a problem, we recommend using the transpiled version for these anyway. However, you do need to fix the Babel issue by installing the plugin it asks for, because right now if I merge this, transpilation is broken (which is why you're seeing the failing check here).
If your code was not working on a recent Chrome version, then there must be a bug with your code, since the object spread operator is supported there. What are you using to test this? I still don't see any functions enabled with the functionality in this PR.

efeichen commented:
Oh yes, I've enabled the metadata on some Mavo functions but haven't committed them yet. After implementing your earlier suggestion everything is working now :)

LeaVerou commented:
Thanks, this looks good! I will merge it after we release v0.2.0 in a few days since I'm avoiding major changes until then.

Question: Why was `default: null` needed? Isn't that the default default anyway?

efeichen commented:
I will clean up the tests for this and push them soon. Also, this PR has changed pretty much everything PR #374 originally was intending to change, except the loop that wraps unary Math functions. Should I close 374 and open up new PRs for future changes on this?

efeichen commented:
It doesn't look like all Math functions are the properties of `Mavo.Functions` only the explicitly defined/ overloaded ones. Is there a way to access *all* functions including Math ones in Mavo?

LeaVerou commented:
Oh right, they are exposed via the proxy, not by adding them on `Mavo.Functions`. My bad!

Your logic seems correct, just instead of wrapping them in place, wrap them and add them on `Mavo.Functions`, to avoid messing with globals we don't own.

LeaVerou commented:
Or, actually, now that I think about it, it might be easier to just add them to `Mavo.Functions` and set `multiValued` to true, then let the loop to its job.

efeichen commented:
Could you briefly explain what the proxy in `data.js` does? I noticed that after deleting [this](https://github.com/mavoweb/mavo/blob/490d713daabf0fae1963b424f7e72f925d7b92b8/src/data.js#L156) line, my test `atan2([0,5], [2,5]) * 180 / PI`, which correctly returned [0, 45], failed.

efeichen commented:
Does the removal of that line also somehow take away the ability to perform element-wise operation with the result of the function and another operand? Something like ` atan2([0,5], [2,5]) * 180 / PI` worked in the past, but after deleting that line, the result becomes `[null, null]`.

I could try investigating myself first and I'll message you if I get stuck. I'd like to explore and sharpen my debugging skill a little bit :)

LeaVerou commented:
No, it should just take away the function itself, but I don't remember how Mavo handles non-existent functions.

efeichen commented:
Ah ya, the reason why it became `null` was because after deleting that line the `PI` constant becomes `undefined`. If this is the case, would you still say removing logic regarding `Math` in `data.js` and explicitly defining them in `functions.js` is a good idea? I have to define 29 more functions. Although I don't mind the chore, it does seem quite verbose.

LeaVerou commented:
If we do it that way it would be done via a loop, not manually!!!

LeaVerou commented:
I changed my mind. Let's get this in to v0.2.0. Merged, thanks!

LeaVerou commented:
Btw I noticed that functions with just 1 argument (e.g. `uppercase()`) are still handled via `binaryOperation` but it somehow seems to work fine. Is that because `binaryOperation` just works fine in that case, or have you done some handling that I missed?

efeichen commented:
Ya `binaryOperation` works fine. And plus, if I check for `fn.length === 1`, some functions with default parameters may be misinterpreted and wrapped with `unaryOperation`, so I removed that logic until perhaps later.

LeaVerou commented:
Thanks!

Btw you may want to take a look at why these `contains()` tests are now (?) failing: https://test.mavo.io/functions.html#contains

efeichen commented:
Right, I noticed that quite a while ago. This is what I'm going to fix next.
